DDF-3439 Adding Error Prone static analysis plugin

coyotesqrl opened the pull request:
#### What does this PR do?
Adds [Google's Error Prone](https://github.com/google/error-prone) static analysis tool through maven. This will run compile-time analysis of files with a modern and active ruleset.

This does of course increase compilation time. On my machine, I am seeing quick builds of ~15 minutes.

#### Who is reviewing it? 
@clockard @pklinef @rzwiefel @stustison @lessarderic 

(please choose AT LEAST two reviewers that need to approve the PR before it can get merged)
#### Select relevant component teams: 
https://github.com/orgs/codice/teams

#### Choose 2 committers to review/merge the PR. 
(please choose ONLY two committers from below, delete the rest)

#### How should this be tested? (List steps with links to updated documentation)
* CI build 
* Builds on local machines to determine if the increase in compilation time is palatable
* Installation and sanity test to affirm that compilation has not been impacted.

#### Any background context you want to provide?
N/A

#### What are the relevant tickets?
[](https://codice.atlassian.net/browse/)

#### Screenshots (if appropriate)
#### Checklist:
- [ ] Documentation Updated
- [X] Update / Add Unit Tests
- [ ] Update / Add Integration Tests
#### Review Comment Legend:
- ✏️ (Pencil) This comment is a nitpick or style suggestion, no action required for approval. This comment should provide a suggestion either as an in line code snippet or a gist. 
- ❓ (Question Mark) This comment is to gain a clearer understanding of design or code choices, clarification is required but action may not be necessary for approval.
- ❗ (Exclamation Mark) This comment is critical and requires clarification or action before approval.

coyotesqrl commented on pom.xml:
```diff
@@ -512,6 +512,36 @@
+                <configuration>
+                    <compilerId>javac-with-errorprone</compilerId>
+                    <compilerArgs>
+                       <arg>-Xep:JUnit4TestNotRun:WARN</arg>
```
I had intended to leave all error checks at their default settings; however, this check proved to be too annoying to resolve. Error Prone expects ignored unit tests to have both the `@Ignore` and the `@Test` annotation. In many of our ignored cases, we only have the `@Ignore` annotation. As the test still does not run and the intent is clear, it seems overkill to update the several test classes in the system that are incorrect in this small way just to satisfy the tool.

coyotesqrl commented:
In addition to resolving the Error Prone blockers, I fixed several SonarQube findings in the classes in the PR; I did not fix _all_ SonarQube findings.

coyotesqrl commented:
build now

bdeining commented on distribution/test/itests/test-itests-common/src/test/java/ddf/common/test/PaxExamRuleTest.java:
```diff
@@ -176,7 +174,7 @@ public void failingBeforeExam() {
             .anyMatch(failure -> failure.getMessage().contains(EXPECTED_BEFORE_EXAM_ERROR_MESSAGE));
 
-    assertThat(examFail && beforeExamFail);
+    assertTrue(examFail && beforeExamFail);
```
This change goes against our unit testing guidelines.  https://codice.atlassian.net/wiki/display/DDF/Unit+Testing

bdeining commented on distribution/test/itests/test-itests-common/src/test/java/ddf/common/test/PaxExamRuleTest.java:
```diff
@@ -197,7 +195,7 @@ public void failingAfterExam() {
             .anyMatch(failure -> failure.getMessage().contains(EXPECTED_AFTER_EXAM_ERROR_MESSAGE));
 
-    assertThat(examFail && afterExamFail);
+    assertTrue(examFail && afterExamFail);
```
This change goes against our unit testing guidelines.  https://codice.atlassian.net/wiki/display/DDF/Unit+Testing

bdeining commented on catalog/spatial/csw/spatial-csw-endpoint/src/test/java/org/codice/ddf/spatial/ogc/csw/catalog/endpoint/reader/TestTransactionMessageBodyReader.java:
```diff
@@ -312,7 +312,7 @@ public void testIsReadable() throws Exception {
   public void testReadInsertFrom() throws Exception {
     Converter mockConverter = mock(Converter.class);
-    when(mockConverter.canConvert(any(Metacard.class.getClass()))).thenReturn(true);
+    when(mockConverter.canConvert(any(Class.class))).thenReturn(true);
```
is this different that just using `any()` without any arguments?

coyotesqrl replied:
Not especially, but I figured I'd leave it explicit.

clockard commented:
Can this be turned off with the `-DskipStatic=true`?

coyotesqrl commented:
It cannot be turned off. It is a compilation step.

coyotesqrl commented:
Just pushed some significant additional changes to this PR:
* Removed FindBugs analysis (note that we still use the `com.google.code.findbugs:jsr305` artifact for resolving `javax.annotation.*` annotations).
* Resolved a swath of SonarQube findings in the affected files, but not all of them at present.
* Changed the way ErrorProne is configured in the reactor pom.

rzwiefel commented on catalog/rest/catalog-rest-endpoint/src/test/java/org/codice/ddf/endpoints/rest/TestRestEndpoint.java:
```diff
@@ -909,7 +909,7 @@ protected String mockTestSetup(CatalogFramework framework, TestType testType)
         transformer = "resource";
         /* FALLTHRU */
-
+        // fall through
```
❓ Did you just add a second fall through comment because the one on line 911 isn't understood by the static analysis?

coyotesqrl replied:
Yes. ErrorProne has an expectation of how the comment should look. I contemplated removing the first comment, but suspect it might be silencing one of our _other_ static analysis tools.

coyotesqrl commented:
build now

coyotesqrl commented:
build now

coyotesqrl commented:
build now

coyotesqrl commented:
build now

coyotesqrl commented on platform/security/sts/security-sts-server/src/main/java/ddf/security/sts/EHCacheTokenStore.java:
```diff
@@ -71,19 +70,23 @@ public EHCacheTokenStore(String key, Bus b, URL configFileURL) {
     Cache newCache = new RefCountCache(cc);
     cache = cacheManager.addCacheIfAbsent(newCache);
-    synchronized (cache) {
+    lock.lock();
```
@roelens8 Would you please review the changes I made to this class? I had to update it for the removal of FindBugs and noted this sonarqube issue while reviewing my changelist. The lock usage is correct now, but I'm not sure it matches the semantics you were going for.

roelens8 replied:
👍

coyotesqrl commented:
Intentionally side-stepped CI to get this static analysis change merged more quickly. Each merge ahead of it had introduced new failures to be addressed; this will protect master more quickly.
